### PR TITLE
Reworking FileModule implementations

### PR DESCRIPTION
**Pull request by travisbrown**
This is a follow-up to @johnynek's #106 that splits up the implementation of `FileModule` depending on whether the context is suspendable or not.

Part of the motivation is that the name `captureEffect` wasn't really accurate for the suspendable case previously, since it just captured thrown exceptions, not _effects_. In the new version the signature includes the `Eval`, so it's a little more accurate.

This split also lets us avoid stepping on each other's toes. We have a single `FileModule` interface, but well-behaved contexts like the `Task`s and `Rerunnable` don't have to pay any tax because of implementations that are shared with non-stack-safe, non-suspendable contexts like `Try`.

I've also introduced new modules for Twitter's `Try` and `Future`, so that instead of the `twitter` package just giving you the `Rerunnable` module, you actually get to choose. We should probably do the same for more of the standard library's contexts, but I'll leave that for a follow-up PR if the general approach here looks agreeable to everyone.

### CODE REVIEW

**Comment by travisbrown** [comment]
@johnynek For what it's worth here are benchmark results for counting the average word length in Bartleby the Scrivener for Scalaz's `Task`, Twitter's `Future`, `Rerunnable`, and `Try`:

```
[info] Benchmark                             Mode  Cnt    Score    Error  Units
[info] FileModuleBenchmark.avgWordLengthS   thrpt   40  372.964 ±  1.785  ops/s
[info] FileModuleBenchmark.avgWordLengthTF  thrpt   40  285.083 ±  1.418  ops/s
[info] FileModuleBenchmark.avgWordLengthTR  thrpt   40  283.991 ±  1.166  ops/s
[info] FileModuleBenchmark.avgWordLengthTT  thrpt   40  584.586 ± 16.037  ops/s
```

File: files/src/main/scala/io/iteratee/files/SuspendableFileModule.scala [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    Enumerator.liftM(captureEffect(new BufferedInputStream(new FileInputStream(file))))(F).flatMap { stream =>
+      new ByteEnumerator(stream).ensure(close(stream))(F)
+    }(F)
+
+  final def readBytesFromStream(stream: InputStream): Enumerator[F, Array[Byte]] =
+    Enumerator.liftM(captureEffect(new BufferedInputStream(stream)))(F).flatMap { stream =>
+      new ByteEnumerator(stream).ensure(close(stream))(F)
+    }(F)
+
+  final def readZipStreams(file: File): Enumerator[F, (ZipEntry, InputStream)] =
```
Comment (johnynek) [inline files/src/main/scala/io/iteratee/files/SuspendableFileModule.scala]: shame we can't share these.

Like a common `protected def catch[A](a: => A): F[A]` and in Suspendable `catch == capture` in `NonSuspendable` `catch == catchNonFatal`

**Comment by johnynek** [comment]
looks good to me. Just a minor comment about a way to share some of the implementations (mainly the ones on `InputStream` I guess, because they are already open, you don't need to capture the effect of opening them).

Interesting about the benchmarks. I'm surprised `Task` performs so much better than `Future` here.

Good to see that `Try` is actually a fair amount faster, so I was not totally wasting my time pushing for all this.

**Comment by travisbrown** [comment]
I agree it's a shame to have this duplication, but even in the stream cases you've still got `ensureEval`. Since you're okay with the API and general approach, @johnynek, I'm going to go ahead and merge this with the goal of getting the final 0.6.0 published today, and I can look into ways of avoiding the repetition in the implementations later.

File: benchmark/src/main/scala/io/iteratee/benchmark/FreeTryModule.scala [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  type FreeTry[x] = Free[Try, x]
+  final type M[f[_]] = MonadError[f, Throwable]
+
+  def captureEffect[A](a: => A): Free[Try, A] = Free.suspend(Free.liftF(MonadError[Try, Throwable].catchNonFatal(a)))
+
+  final protected val F: MonadError[FreeTry, Throwable] = new MonadError[FreeTry, Throwable] {
+    private[this] val FF = Monad[FreeTry]
+    def pure[A](x: A): Free[Try, A] = FF.pure(x)
+    def raiseError[A](e: Throwable): Free[Try, A] = Free.liftF(MonadError[Try, Throwable].raiseError(e))
+    def handleErrorWith[A](fa: Free[Try, A])(f: Throwable => Free[Try, A]): Free[Try, A] = Free.liftF(
```
Comment (johnynek) [inline benchmark/src/main/scala/io/iteratee/benchmark/FreeTryModule.scala]: Though I don't think you use it, don't you want suspend around the value you have returned from handleErrorWith? Now it is strict, no?

**Pull request merged**